functions/go/set-namespace/namespace_transformer.go

```diff
@@ -60,6 +71,67 @@ func (p *plugin) Transform(m resmap.ResMap) error {
 		if err != nil {
 			return err
 		}
+		if err = p.roleBindingHack(r, sAL); err != nil {
+			return err
+		}
 	}
 	return nil
 }
+
+// roleBindingHack extends the behavior of the upstream kustomize filter
+// This extension adds the following behavior:
+// Given a ServiceAccount is present in the ResourceList, additionally set the
+// namespace for that ServiceAccount where it is referenced in the "subjects"
+// field of a RoleBinding or ClusterRoleBinding
+func (p *plugin) roleBindingHack(r *resource.Resource, sAL ServiceAccountLookup) error {
+	kind := r.GetKind()
+	if kind != roleBindingKind && kind != clusterRoleBindingKind {
+		return nil
+	}
+	subjects, err := r.Pipe(kyaml.Lookup(subjectsField))
+	if err != nil || kyaml.IsMissingOrNull(subjects) {
+		return err
+	}
+
+	err = subjects.VisitElements(func(o *kyaml.RNode) error {
+		subjectKind := o.GetKind()
+		if subjectKind != serviceAccountKind {
+			return nil
+		}
+		var nameRN *kyaml.RNode
+		nameRN, err = o.Pipe(kyaml.Lookup("name"))
+		if err != nil || kyaml.IsMissingOrNull(nameRN) {
+			return err
+		}
+		nameStr := kyaml.GetValue(nameRN)
+		if !sAL.HasServiceAccount(nameStr) {
+			return nil
+		}
+		v := kyaml.NewScalarRNode(p.Namespace)
+		return o.PipeE(
+			kyaml.LookupCreate(kyaml.ScalarNode, "namespace"),
+			kyaml.FieldSetter{Value: v},
+		)
+	})
+	return err
+}
+
+// ServiceAccountLookup provides an API for tracking ServiceAccount resources
+type ServiceAccountLookup struct {
```

Comment:
This struct doesn't need to be exported.

Reply:
I was aiming to present this API in a way that could be easily migrated over to the shared library (kustomize), assuming this behavior is well received. In which case, assuming the API remains unchanged, would mean the struct would move to the shared library and be exported there.

However, it's a fair point that it does not need to be exported in this iteration - which is probably sufficient justification to not export it. I shall un-export it 🙂